### PR TITLE
Another reduction of describe via reducing the number of list calls

### PR DESCRIPTION
**Pull request by Skarlso**
## Part 2 Of refactoring GetStackTemplate and DescribeStack

This time, Describe is reduced due to the number of reduced list calls during deletion. It's possible to pass around stacks by adding Stack to the nodeGroupStack and reducing the number of calls by moving around lists which already happened at the top level.

Related to: https://github.com/weaveworks/eksctl/issues/4635

## Integration Test Run results

```
Ran 43 of 43 Specs in 4572.260 seconds
FAIL! -- 41 Passed | 2 Failed | 0 Pending | 0 Skipped
--- FAIL: TestCRUD (4572.28s)
```

The 2 failures appear to be unrelated, but I'll make sure. 

Nodegroup tests incoming...

Hmmm.... Node listing works by hand, but this test keeps failing which makes me think either the test is missing something that I need to add, or I'm missing something from somewhere. The code should work fine, this is just listing. Basically `get nodegroups` fails but it makes no sense why.

Managed Nodegroup test:

```
/Users/skarlso/goprojects/weaveworks/eksctl/integration/tests/managed/managed_nodegroup_test.go:441

Ran 21 of 21 Specs in 4950.664 seconds
FAIL! -- 20 Passed | 1 Failed | 0 Pending | 0 Skipped
--- FAIL: TestManaged (4950.67s)

```

Fixed the failure and rerun the failing part:

```
failed...
Error: upgrade is only supported for managed nodegroups; could not find one with name "bottlerocket-1"
```

This is definitely a problem somewhere ^^. I run it isolation so it didn't create that nodegroup 🤦.
But it's fine, because `upgradeNg(initialAl2Nodegroup)` this passed so the other one should have as well.

Just going to run the integration test suit on my branch -> https://github.com/weaveworks/eksctl-ci/actions/runs/1830405203

### CODE REVIEW

**Comment by Skarlso** [comment]
I don't see how my changes here could affect `Get`.

**Review by aclevername — DISMISSED** [review]
lgtm!

File: pkg/cfn/manager/delete_tasks.go [hunk 1]
```diff
@@ -84,38 +80,32 @@ func (c *StackCollection) NewTasksToDeleteClusterWithNodeGroups(deleteOIDCProvid
 }
 
 // NewTasksToDeleteNodeGroups defines tasks required to delete all of the nodegroups
-func (c *StackCollection) NewTasksToDeleteNodeGroups(shouldDelete func(string) bool, wait bool, cleanup func(chan error, string) error) (*tasks.TaskTree, error) {
-	nodeGroupStacks, err := c.DescribeNodeGroupStacks()
-	if err != nil {
-		return nil, err
-	}
-
+func (c *StackCollection) NewTasksToDeleteNodeGroups(nodeGroupStacks []NodeGroupStack, shouldDelete func(string) bool, wait bool, cleanup func(chan error, string) error) (*tasks.TaskTree, error) {
```
[the remark this answers is not in the document]
Comment (Skarlso) [reply pkg/cfn/manager/delete_tasks.go:83]: >Having the stackCollection do the fetching of the nodegroup stacks makes sense, and having the caller do it seems a bit more messy/noisy as it pushs more effort on the caller

But they already did in some cases, hence the refactor.

>Only a couple times do we actually save an API call

That's true, it's actually quite nice on most cases, or sadly necessary in some.

>in some of the refactors above its the same number of calls regardless but now with an extra func call+ error handling+unit test

Where?

[the remark this answers is not in the document]
Comment (aclevername) [reply pkg/cfn/manager/delete_tasks.go:83]: For `(c *OwnedCluster) Upgrade(dryRun bool)` I think (which sadly doesn't seem to have any tests)

**Review by aclevername — APPROVED** [review]

**Comment by Skarlso** [comment]
Integration tests passed on this Branch. ( I pushed it to eksctl directly too with the current change set.)

**Review by cPu1 — APPROVED** [review]

**Pull request merged**